src/Templates/SingleProductTemplateCompatibility.php

```diff
@@ -0,0 +1,425 @@
+				'position'   => 'before',
+				'hooked'     => array(
+					'woocommerce_output_product_data_tabs' => 10,
+					'woocommerce_upsell_display'           => 15,
+					'woocommerce_output_related_products'  => 20,
+				),
+			),
+		);
+	}
+
+	/**
+	 * Add compatibility layer to the first and last block of the Single Product Template.
+	 *
+	 * @param string $template_content Template.
+	 * @return string
+	 */
+	public static function add_compatibility_layer( $template_content ) {
+		$wrapped_blocks = self::wrap_single_product_template( $template_content );
+		$template       = self::inject_custom_attributes_to_first_and_last_block_single_product_template( $wrapped_blocks );
+
+		return array_reduce(
+			$template,
+			function( $carry, $item ) {
+				if ( is_array( $item ) ) {
+					return $carry . serialize_blocks( $item );
+				}
+				return $carry . serialize_block( $item );
+			},
+			''
+		);
+
+	}
+
+	/**
+	 * For compatibility reason, we need to wrap the Single Product template in a div with specific class.
+	 * For more details, see https://github.com/woocommerce/woocommerce-blocks/issues/8314.
+	 *
+	 * @param string $template_content Template Content.
+	 * @return array Wrapped template content inside a div.
+	 */
+	private static function wrap_single_product_template( $template_content ) {
+		$parsed_blocks  = parse_blocks( $template_content );
+		$grouped_blocks = self::group_blocks( $parsed_blocks );
+
+		// @TODO: Check this list before terminating the Blockfied Single Product Template project.
+		$single_product_template_blocks = array( 'woocommerce/product-image-gallery', 'woocommerce/product-details', 'woocommerce/add-to-cart-form' );
+
+		$wrapped_blocks = array_map(
+			function( $blocks ) use ( $single_product_template_blocks ) {
+				if ( 'core/template-part' === $blocks[0]['blockName'] ) {
+					return $blocks;
+				}
+
+				$has_single_product_template_blocks = self::has_single_product_template_blocks( $blocks, $single_product_template_blocks );
+
+				if ( $has_single_product_template_blocks ) {
+					$wrapped_block = self::create_wrap_block_group( $blocks );
+					return array( $wrapped_block[0] );
+				}
+				return $blocks;
+			},
+			$grouped_blocks
+		);
+		return $wrapped_blocks;
+	}
+
+	/**
+	 * Add custom attributes to the first group block and last group block that wrap Single Product Template blocks.
+	 *
+	 * @param array $wrapped_blocks Wrapped blocks.
+	 * @return array
+	 */
+	private static function inject_custom_attributes_to_first_and_last_block_single_product_template( $wrapped_blocks ) {
+		$template_with_custom_attributes = array_reduce(
+			$wrapped_blocks,
+			function( $carry, $item ) {
+
+				$index          = $carry['index'];
+				$carry['index'] = $carry['index'] + 1;
+				$block          = $item[0];
+
+				if ( 'core/template-part' === $block['blockName'] ) {
+					$carry['template'][] = $block;
+					return $carry;
+				}
+
+				if ( false === $carry['first_block']['found'] ) {
```

Comment:
Not a blocker at all, but I think the code could be simplified if we removed the `found` attribute and instead checked whether `index` is not empty.